src/classes/BDI_DataImportService.cls

```diff
@@ -1725,66 +1725,81 @@ global with sharing class BDI_DataImportService {
     */
     private void importCampaigns() {
 
-        // since we don't currently fill in a campaign lookup or status on the DI record,
-        // there is no work to do in Dry Run.
-        if (isDryRun) {
-            return;
-        }
-
         // need to find all the Campaigns that we might need
-        Set<String> setStrCampaign = new Set<String>();
+        Set<String> uniqueCampaignNames = new Set<String>();
+        Set<Id> setCampaignId = new Set<Id>();
+
         for (DataImport__c dataImport : listDI) {
             // skip di's that already have an error
             if (dataImport.Status__c == BDI_DataImport_API.bdiFailed) {
                 continue;
             }
 
             if (dataImport.Donation_Campaign_Name__c != null) {
-                setStrCampaign.add(dataImport.Donation_Campaign_Name__c);
+                uniqueCampaignNames.add(dataImport.Donation_Campaign_Name__c);
             }
 
         }
 
         // add the existing campaigns to our map
-        for (Campaign cmp : [SELECT Id, Name FROM Campaign WHERE Name IN :setStrCampaign]) {
-            mapStrNameToCampaignId.put(cmp.Name, cmp.Id);
+        for (Campaign cmp : [SELECT Id, Name FROM Campaign WHERE Name IN :uniqueCampaignNames]) {
+            campaignNameToCampaignId.put(cmp.Name, cmp.Id);
         }
 
         // create any new campaigns
-        List<Campaign> listCmpNew = new List<Campaign>();
-        List<DataImport__c> listDINew = new List<DataImport__c>();
-        setStrCampaign.clear();
+        List<Campaign> campaignsToInsert = new List<Campaign>();
+        List<DataImport__c> dataImportsWithNewCampaigns = new List<DataImport__c>();
+        uniqueCampaignNames.clear();
         for (DataImport__c dataImport : listDI) {
             // skip di's that already have an error
             if (dataImport.Status__c == BDI_DataImport_API.bdiFailed) {
                 continue;
             }
 
-            String strName = dataImport.Donation_Campaign_Name__c;
+            String campaignName = dataImport.Donation_Campaign_Name__c;
+            Id campaignId = dataImport.DonationCampaignImported__c;
             Campaign cmp;
-            if (strName != null && setStrCampaign.add(strName)) {
-                if (mapStrNameToCampaignId.get(strName) == null) {
-                    cmp = new Campaign(Name=strName, IsActive=true);
-                    listCmpNew.add(cmp);
-                    listDINew.add(dataImport);
+
+            if (campaignId != null) {
+                // User specified a Campaign lookup in BGE lookup or with an ID on import
+                dataImport.DonationCampaignImportStatus__c = System.Label.bdiMatchedId;
+            } else if (campaignName != null && uniqueCampaignNames.add(campaignName)) {
```

Comment:
Just curious - why the uniqueCampaignNames.add(campaignName) here?  I don't see any subsequent usage of the uniqueCampaignNames Set.

Reply:
Great question, and I actually removed this at one point because it seemed unnecessary and it broke a test! Since uniqueCampaignNames is a set, it returns a boolean that lets us know if it could be added and wasn't in the list yet, or if it couldn't be added because we've already queued up a campaign with that same name. This stops us from making duplicate campaigns.

Reply:
Would it be clearer to separate the addition from the check? 

```
else if (campaignName != null && !uniqueCampaignNames.contains(campaignName)) {
    uniqueCampaignNames.add(campaignName)
```

Reply:
Yeah, I think so @lmeerkatz. A little more code, but definitely easier to read.